controllers/helmchart_controller.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2020 The Flux CD contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/repo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1alpha1"
+)
+
+// HelmChartReconciler reconciles a HelmChart object
+type HelmChartReconciler struct {
+	client.Client
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+	Storage *Storage
+	Kind    string
+	Getters getter.Providers
+}
+
+// +kubebuilder:rbac:groups=source.fluxcd.io,resources=helmcharts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=source.fluxcd.io,resources=helmcharts/status,verbs=get;update;patch
+
+func (r *HelmChartReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+
+	log := r.Log.WithValues("helmchart", req.NamespacedName)
+
+	var chart sourcev1.HelmChart
+	if err := r.Get(ctx, req.NamespacedName, &chart); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// set initial status
+	if reset, status := r.shouldResetStatus(chart); reset {
+		log.Info("Initializing chart")
+		chart.Status = status
+		if err := r.Status().Update(ctx, &chart); err != nil {
+			log.Error(err, "unable to update HelmChart status")
+			return ctrl.Result{}, err
+		}
+	}
+
+	// try to remove old artifacts
+	r.gc(chart)
+
+	repository, err := r.chartRepository(ctx, chart)
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
We want to requeue the chart if the HelmRepository is not present. (use `return result, err`)

Reply:
```go
ctrl.Result{Requeue: true}
```

causes the resource to be requeued with an exponential back-off by default, which is now used.